fastjet/internal/delaunay/triangulation.go

```diff
@@ -0,0 +1,1066 @@
+// Copyright 2017 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package delaunay
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+)
+
+// Delaunay holds necessary information for the
+// delaunay triangulation
+type Delaunay struct {
+	// triangles is a slice of all triangles that have been created. It is used to get the final
+	// list of triangles in the delaunay triangulation
+	triangles triangles
+	// root is a triangle that contains all points. It is used as the starting point in the hierarchy
+	// to find the triangle that contains a point
+	root                   *Triangle
+	maxX, minX, maxY, minY float64
+	// hierarchy indicates which method to use to find the triangle that contains the point
+	hierarchy bool
+}
+
+// HierarchicalDelaunay creates a delaunay triangulation with the given points
+// all points have to be inside the user defined bounds
+// uses a hierarchy to find the triangle which contains a point
+// has a worst time complexity of O(nln(n))
+func HierarchicalDelaunay(points []*Point, maxX float64, minX float64, maxY float64, minY float64) *Delaunay {
+	// root Triangle is a triangle that contains all points
+	dx := maxX - minX
+	dy := maxY - minY
+	a := NewPoint(minX-8*dx, maxY+10, -1)
+	b := NewPoint(maxX+8*dx, maxY+10, -1)
+	c := NewPoint(minX+dx/2, minY-8*dy, -1)
+	root := NewTriangle(a, b, c)
+	d := &Delaunay{
+		root:      root,
+		maxX:      maxX,
+		minX:      minX,
+		maxY:      maxY,
+		minY:      minY,
+		hierarchy: true,
+	}
+	d.triangles = make([]*Triangle, 0)
+	for _, p := range points {
+		d.InsertPoint(p)
+	}
+	return d
+}
+
+// WalkDelaunay creates a delaunay triangulation with the given points
+// it uses the remembering stochastic walk method to find the triangle in which p is inserted
+// it has a worst time complexity of O(n^5/3) but it doesn't need bounds
+func WalkDelaunay(points []*Point) *Delaunay {
+	// it needs at least one triangle to start with
+	if len(points) < 3 {
+		panic(fmt.Errorf("delaunay: not enough points"))
+	}
+	rand.Seed(time.Now().UnixNano())
```

Comment:
Take a `*rand.Rand` parameter so that the user can decide what the seed is. If it is `nil` use `rand.Intn`, but a library should not contaminate global variable behind a user's back.